les/servingqueue.go

```diff
@@ -170,17 +193,119 @@ func (sq *servingQueue) threadController() {
 	}
 }
 
+type (
+	// peerTasks lists the tasks received from a given peer when selecting peers to freeze
+	peerTasks struct {
+		peer          *peer
+		list          []*servingTask
+		sumTime       uint64
+		worstPriority int64
+	}
+	// peerList is a sortable list of peerTasks
+	peerList []*peerTasks
+)
+
+func (l peerList) Len() int {
+	return len(l)
+}
+
+func (l peerList) Less(i, j int) bool {
+	return (l[i].worstPriority - l[j].worstPriority) < 0
+}
+
+func (l peerList) Swap(i, j int) {
+	l[i], l[j] = l[j], l[i]
+}
+
+// freezePeers selects the peers with the worst priority queued tasks and freezes
+// them until burstTime goes under burstDropLimit or all peers are frozen
+func (sq *servingQueue) freezePeers() {
+	peerMap := make(map[*peer]*peerTasks)
+	var peerList peerList
+	if sq.best != nil {
+		sq.queue.Push(sq.best, sq.best.priority)
+	}
+	sq.best = nil
+	for sq.queue.Size() > 0 {
+		task := sq.queue.PopItem().(*servingTask)
+		tasks := peerMap[task.peer]
+		if tasks == nil {
+			tasks = &peerTasks{
+				peer:          task.peer,
+				worstPriority: task.priority,
+			}
+			peerMap[task.peer] = tasks
+			peerList = append(peerList, tasks)
+		} else {
+			if tasks.worstPriority-task.priority > 0 {
+				tasks.worstPriority = task.priority
+			}
+		}
+		tasks.list = append(tasks.list, task)
+		tasks.sumTime += task.expTime
+	}
+	sort.Sort(peerList)
+	drop := true
+	sq.logger.Event("freezing peers")
+	for _, tasks := range peerList {
```

Comment:
The freezing strategy here is: selects the peers with the worst priority queued tasks. It is kind of unfair.

The priority is decided by two factors: (1) when the task is created (2) whether the requesting peer is recharging. 

So if a priority peer sends a new request, the worst priority could less than a free client which sends a lot of requests.

Reply:
I think we should freeze free client first and then freeze priority client which sends more requests(the recharge needed time is longer).

Reply:
I think you are right, the time factor is not really relevant in this case. Now the peers are sorted based on relative buffer status (bufValue/bufLimit), exhausted buffers are frozen first.

Reply:
About freezing free clients first: I did think about adding an extra condition to favorize priority peers even more but I am not sure that would be desirable. The main point of this capacity model is that service quality is controlled by a simple scalar value for all peers. If you are a priority peer and pay for a high capacity then you will of course have a better relative buffer status after using the same amount of requests because your bufLimit is higher. Still, priority peers also need to be throttled eventually. From a technical point of view, if the problem is too many requests at the moment then you can only solve that by freezing heavy requesters.

Reply:
Yes you are right, it makes more sense here to use `relative buffer status` as the priority.